re-organise the compiler team (#3599)
rust-lang/rfcs: text/0000-compiler-team-trusted-contrib-maintainer.md

compiler team contributor is a role within the project which recognises regular
contribution and trust rather than just having made a handful of contributions
and thus being a contributor.

[rfc2689]: https://rust-lang.github.io/rfcs/2689-compiler-team-contributors.html
[review_queue_analysis]: https://borrowed.dev/p/on-the-compiler-teams-review-queue

Definitions
-----------
[definitions]: #definitions

There are various permissions/privileges/responsibilities which will be
referenced in later sections of this RFC, defined here:

- **r+**
  - Contributors with *r+* privilege are able to approve pull requests to be
    merged by *bors*. Contributors should not merge their own pull requests (with
    the exception of re-approving their own work on behalf of another contributor
    after a rebase or similarly trivial change). *r+* permissions apply to the
    whole repository, but [it is expected][expectations] that contributors limit
    themselves to only those parts of the *rust-lang/rust* repository that are under
    the purview of the compiler team (unless granted *r+* from other teams too), and
    for subsystems/pull requests that they are confident reviewing.

> - **try**

Reviewer: question that I'm gonna answer myself: why don't just give everyone permissions for try builds? the impact of the other perms is obvious, but it's notable that try permissions are a security risk as try builds have access to secrets, and they're also hosted on static.rust-lang.org and therefore must never contain malicious code

Author: Added some wording to this effect, thanks!

  - Contributors with *try* permissions are able to trigger complete toolchain
    builds for a pull request or commit, which are then used by *rustc-perf* and
    *crater*.

> - **review queue**

Reviewer 2: This is usually (and even later in this document) called "Review rotation"

Author: Fixed!

    group mentions and receive a "Member" badge next to their name.
- **rustc-perf**
  - Contributors with permissions to use *rustc-perf* can request benchmarking
    of their pull requests (and pull requests they are reviewing). *rustc-perf*
    permissions are useful for regular contributors as it is common to need to
    request benchmarks from contributors with permissions. *rustc-perf* permissions
    only make sense alongside *try* permissions.
- **crater**
  - Contributors with permissions to *crater* can request crater runs to check
    whether their code breaks any public ecosystem code. *crater* permissions only
    make sense alongside *try* permissions.
- **dev desktops**
  - Contributors with access to developer desktops are able to connect to shared
    development servers that they can do their contributions from.

[expectations]: https://forge.rust-lang.org/compiler/reviews.html#expectations-for-r

Guide-level explanation
=======================
[guide-level-explanation]: #guide-level-explanation

Contributors start without any particular privileges, permissions or
responsibilities and can contribute whatever they'd like. Contributors can

> progress to [Trusted Contributors][trusted-contributors] and then [Team Members]

Reviewer 3: I've very much warmed up to the idea of changing the naming to "contributor" -> "member" (from trusted contributor) -> "maintainer" (from member).

First, the "maintainer" name does seem to represent better what these members (participants? we somewhat overload this word in English) are tasked to do.

For "Member" vs contributor, see the language: What does someone interested who submits a couple PRs to the rust repo do? They *contribute*. If we give them permissions for bors and such, and do various things, they're now a *member* of whatever team that is.

Ultimately though, its just naming. And the name does differ quite a bit from the current naming. But I think it better matches what we expect when we use these words in everyday English.

Author: I agree, this aligns with my own thinking on the topic more too. My only concern is that it diverges from what other teams have a little bit. I've renamed these in the RFC.

> [team-members].

Author: Oops, fixed!

nominate a contributor. Team leads will check for a reasonable contribution
history, and will check if the current team have any serious concerns related to
contributor conduct (waiting approximately one week).

When evaluating a candidate's contribution history, length of time and
consistency of contributions and interactions with other contributors and team
members will be taken into account. It is important to note that many kinds
of contributions will be considered such as code contributions, helping with
issue triage and bisection, running meetings and creating minutes, documentation
contributions for rustc internals or the [Compiler Development Guide], etc.

Trusted contributor is a mix of RFC 2689's "working group participant" and
"compiler team contributor" roles. It is explicitly intended to be granted more
liberally to contributors who have demonstrated competence and trustworthiness,
for whom they would be able to work more effectively with these permissions and
can be trusted to use them responsibily. Trusted contributors do not need to
have experience with most of the compiler, and can be specialised to specific
subsystems of the compiler.

> Trusted contributors are granted *r+*, *try*, *triagebot*, *rustc-perf*, and

Author: Added!

*crater* permissions; *organisation membership*; and *dev desktop* access.
Trusted contributors are considered members of the Rust project as a whole, and
are automatically eligible for any benefits that incurs (e.g. invitations to
meetups of project members). As representatives of the Rust project, trusted
contributors are expected to obey not just the letter of the [Code of Conduct]

> [coc] but its spirit.

Reviewer: this link isn't rendered correctly

Author: Oops, fixed!

desire.

[Compiler Development Guide]: https://rustc-dev-guide.rust-lang.org/

Team Members
------------
[team-members]: #team-members

Trusted contributors are eligible to become team members after they have
continued to contribute actively for a year. Trusted contributors can contact
team leads or will be contacted by team leads to enquire about promotion to team
membership. Trusted contributors who are eligible for team membership do not
have to become team members.

Unlike trusted contributors, team members are expected to consider themselves
as *maintainers* of the compiler - put otherwise, to be invested in the quality
of the compiler codebase and overall health of the compiler team, independent
of their own projects. Team membership is primarily intended to recognise and
encourage participation in activities which are vital to the success of the
compiler team and broader project.

Team members are expected to participate in the ongoing maintainance tasks
that the compiler team is responsible for (with all of the expected caveats
for vacation time, mental health breaks, etc) - listed as [responsibilities]

> [responsibilities] below. However, not all contributors need to participate

Author: Oops, fixed!

in these responsibilities to an equal degree. Contributors should participate
in these tasks to the degree that they are able - volunteers are not expected
to participate as much as those employed to work on the compiler, for example.
It is the responsibility of the compiler team leads to ensure that the ongoing
maintenance tasks of the team can be completed sustainably.

Team members have all of the same permissions and access as trusted
contributors. Like trusted contributors, team members are considered members of
the Rust project as a whole and are expected to follow the spirit of the [Code
of Conduct][coc].

Like trusted contributors, after inactivity for longer than a year, a
contributor will be moved to alumni status. Members who are no longer able to
help maintain the compiler but otherwise wish to continue contributing to the
compiler can also be moved to alumni status and retain their trusted contributor

> status. Alumni can ask to be reinstated in future.

Author: Oops, fixed!

[coc]: https://www.rust-lang.org/policies/code-of-conduct

Responsibilities
----------------
[responsibilities]: #responsibilities

There are various responsibilities that a team member could take on to help the
team. All team members must participate in at least one activity.

Team members can get involved in any of these by contacting the team leads, by
asking team members currently involved in these responsibilities, or by asking
in any venue where these responsibilities are conducted (e.g. a Zulip stream).

> - Final comment period (FCP) reviewer

Reviewer 4: I think it would be good to explicitly state some kind of "its expected that fcps are responded to by reviewers within Some Concrete Timeframe" regardless of whether that response is ticking the box or registering a concern or just "im still taking time reviewing this in depth". it tends to Feel Bad when you go onto an FCP and its been open for a while with seemingly no response from someone regardless of whether there's actually a good reason behind the scenes.

Author: I've added wording to this effect.

    - Final comment periods are the process by which the team signs-off on a
      change before it is made, like stabilizing a feature.

      FCPs have always required whole team to sign-off, but this doesn't scale
      as the team grows. As described above, this acts as a disincentive for the
      team to grow. Furthermore, not all FCPs are relevant to all team members and a
      diffusion of responsibility means that most team members just sanity-check and
      then sign-off. This isn't ideal, as it doesn't guarantee that someone on the
      team has thoroughly considered a FCP.

>       Instead, have FCPs require sign-off from team members who opt-in to being

Reviewer 2: I'll note the @rfcbot already supports opt-*out* of being on FCPs. This is [undocumented](https://github.com/rust-lang/team/pull/1419), no one currently uses this and probably not a lot of people know it's an option at all (Mara told me and Boxy about it a couple days ago and both me and Boxy were in complete shock).

But the option does exist and should even work. And it seems that opt-out-ness is deliberate (so it's harder to exclude someone accidentally)...

Reviewer 2: I agree

      an "FCP reviewer", with the expectation that they will spend time reviewing
      an FCP thoroughly. FCP reviewers should also consider reaching out to relevant
      domain experts and soliciting their opinions whenever possible. Any team or
      project member can raise concerns with an FCP, which will be considered by the
      FCP reviewers.
      

>       To function effectively, it is recommended that there be 4 - 8 FCP

Reviewer 5: By what mechanism do we prevent growth beyond the desired range? Like, what if someone wants to be on FCP duty, but we already got 8 folk on the list?

Reviewer 4: Does a 4-8 limit actually matter in practice if all the members are responding "timely". We'd still have the problem of "FCPs span a wide variety of required knowledges and we dont want to encourage mindlessly checking a box because its not your area" but honestly maybe that's fine and if we are explicitly sayign to Not Do This we'd encourage growing the amount of documentation we have for areas so that people can more accurately evaluate fcps

Reviewer 5: "by signing up for FCPs you opt-in to start getting pinged after a week if you haven't reacted to the FCP. Contributors are encouraged to ping FCP folk if there are no more outstanding concerns."

Author: I've added some wording here - we don't want a hard upper limit, the goal is just to have a reasonable amount of diversity of opinion in the FCP list; make sure that everything gets a thorough look; and doesn't take too long. We want the FCP list to be populated by people who are committing to never mindlessly checking their box, they're on the list because they're interested in engaging with all of the FCPs we get. Team members who aren't on the FCP list and have an opinion can still register concerns and things like that to be heard.

Author: > "by signing up for FCPs you opt-in to start getting pinged after a week if you haven't reacted to the FCP. Contributors are encouraged to ping FCP folk if there are no more outstanding concerns."

GitHub hadn't shown me this comment when I was posting mine, I can change the wording to be closer to this if that's preferred.

Reviewer 6: I think specifying expectations for how frequently you're going to be pinged for reviews if you are an FCP reviewer is a good idea!

As David said, the upper limit is really just a suggestion, after ~8 people just collecting reviews from everyone in the group tends to take a while even when no major issues are raised. It's certainly possible that a larger group could also work if the FCP reviewers respond in a very timely manner.

Author: I think the issue wouldn't be that there are too many people, but that some people on the list aren't reviewing promptly.

If we had twenty people on the list and everyone reviewed within a day, I think that would be okay, it's about making sure things move forward and setting expectations about promptness in FCP participation, rather than the number of people (unless maybe it meant that there was always _some_ objection so nothing would progress - but that's a separate problem that we can deal with if we run into it).

      reviewers at any time. If less than 4 FCP reviewers are available, the compiler
      team co-leads will act as FCP reviewers until the reviewers can be found.

- Performance triage
    - There is a rotation of team members and other project members who check
      all of the interesting performance benchmarks from the last week to produce a
      report summarizing the improvements and regressions. This is valuable to keep
      track of the compiler's performance over time and make sure that regressions are
      being addressed.

- Issue prioritisation
    - The compiler team has a prioritisation procedure and policy to identify
      and label issues according to their importance. These labels feed into the
      backport procedure (what's worth being backported) and work priorities of team
      members.

- Backport reviews

>     - Each week, some team members participate in the weekly triage meeting

Reviewer: we've talked about this before but backport decisions should really be changed to be async, so I think it makes sense to be less precise about them on RFC and just say that this means "regularly" deciding on backports

Author: This makes sense, fixed!

      described here will take too long.

Prior art
=========
[prior-art]: #prior-art

- [Responsibilities][responsibilities] are similar to [an unsubmitted proposal
  by Niko Matsakis in December 2020 to have "elected officers"][officers] within
  the compiler team responsible for different team functions. This RFC shares
  many of the goals of Niko's earlier proposal, but is slightly less formal -
  responsibilities are loosely-defined groups of contributors rather than elected
  positions, and there is no rotations or term limits.

  In this RFC's proposal, it is expected that responsibilities are shared
  amongst a group of team members, and that team members do less of other
  responsibilities so that their workload is sustainable, but this isn't enforced.
  Team leads are instead responsible for ensuring that the team is large enough to
  perform each responsibility sustainably.

[officers]: https://hackmd.io/S9xqmwJbSI-a9mFdK9yQBA

Unresolved questions
====================
[unresolved-questions]: #unresolved-questions

None!

Future possibilities
====================
[future-possibilities]: #future-possibilities

> Responsibilities could be formalized further - see references in [Prior art]
> [prior-art].

Reviewer 2: ```suggestion
Responsibilities could be formalized further - see references in [Prior art][prior-art].
```
```suggestion
Responsibilities could be formalized further - see references in
[Prior art][prior-art].
```
💔 🔗

Author: Fixed!

- Feature Name: N/A
- Start Date: 2024-01-19
- RFC PR: [rust-lang/rfcs#3599](https://github.com/rust-lang/rfcs/pull/3599)
- Rust Issue: N/A

Summary
=======
[summary]: #summary

Re-organise the compiler team:

- Re-define and rename the tiers of membership
- Change how team members and contributors are promoted
- Document expectations of team members
- Establish mechanism for scaling additional responsibilities that team members
  take on and recognising these contributions

Motivation
==========
[motivation]: #motivation

Compiler team contributors were introduced [in 2019 with RFC 2689][rfc2689],
the last significant change to the compiler team's structure. A lot has changed

> in the project and compiler team since that time: we receive [approximiately

Reviewer 7: ```suggestion
in the project and compiler team since that time: we receive [approximately
```

Author: Fixed!

    the purview of the compiler team (unless granted *r+* from other teams too), and
    for subsystems/pull requests that they are confident reviewing.
- **try**
  - Contributors with *try* permissions are able to trigger complete toolchain
    builds for a pull request or commit, which are then used by *rustc-perf* and
    *crater*. *try* permissions aren't available to everyone because try builds
    can pose a security risk: try builds have access to secrets and the resulting
    builds are hosted on `static.rust-lang.org` where we would never want
    malicious code.
- **review rotation**
  - Contributors on the review rotation will be randomly assigned to new pull
    requests submitted to the compiler. Being on the review rotation is one of the
    best ways for contributors to help the compiler team and learn new parts of
    the compiler. Review capacity is one of the most important resources that the
    team has, as it enables our progress in the compiler's continued development
    and maintenance.
- **organization membership**
  - Contributors that are added to the *rust-lang/compiler* team in the GitHub

>     organisation can be assigned to issues/pull requests, modify labels, receive

Reviewer 7: Nit: as British spelling is more common in the document, maybe unify the `organization` spelling in the header with the `organisation` one in the body?

Author: Fixed! I've made it the American spelling and that should be what the whole document is, but sometimes I can't help the correct spelling slipping in, so if you spot any others, let me know :)

Reviewer 7: I believe they prefer `z`s where you lovely lot prefer `s`s, e.g. for `Re-organise`, `recognise`, `prioritisation`, and the likes in the document.

    group mentions and receive a "Member" badge next to their name.
- **rustc-perf**
  - Contributors with permissions to use *rustc-perf* can request benchmarking
    of their pull requests (and pull requests they are reviewing). *rustc-perf*
    permissions are useful for regular contributors as it is common to need to
    request benchmarks from contributors with permissions. *rustc-perf* permissions
    only make sense alongside *try* permissions.
- **crater**
  - Contributors with permissions to *crater* can request crater runs to check
    whether their code breaks any public ecosystem code. *crater* permissions only
    make sense alongside *try* permissions.
- **dev desktops**
  - Contributors with access to developer desktops are able to connect to shared
    development servers that they can do their contributions from.
- **triagebot**
  - [triagebot][triagebot] is a GitHub bot that can perform helpful tasks on issues
    and pull requests. Many of its functions are available to everyone, such as issue

>     claiming, but some functions may restricted to project/team members.

Author: Fixed!

work of their collaborators.

Any contributor can request to become a trusted contributor by contacting the
compiler team's leads, or current team members and trusted contributors can
nominate a contributor. Team leads will check for a reasonable contribution
history, and will check if the current team have any serious concerns related to
contributor conduct (waiting approximately one week).

When evaluating a candidate's contribution history, length of time and
consistency of contributions and interactions with other contributors and team
members will be taken into account. It is important to note that many kinds
of contributions will be considered such as code contributions, helping with
issue triage and bisection, running meetings and creating minutes, documentation
contributions for rustc internals or the [Compiler Development Guide], etc.

Trusted contributor is a mix of RFC 2689's "working group participant" and
"compiler team contributor" roles. It is explicitly intended to be granted more
liberally to contributors who have demonstrated competence and trustworthiness,
for whom they would be able to work more effectively with these permissions and

> can be trusted to use them responsibily. Trusted contributors do not need to

Reviewer 8: Nothing escapes the eagle eye of [typos](https://github.com/crate-ci/typos) :)

```suggestion
can be trusted to use them responsibly. Trusted contributors do not need to
```

Author: Fixed!

and other interactions with the project cease) for longer than a year, the
trusted contributor will be moved into alumni status. At any point in the
future, they can ask to be re-instated at the trusted contributor level if they
desire.

[Compiler Development Guide]: https://rustc-dev-guide.rust-lang.org/

Team Members
------------
[team-members]: #team-members

Trusted contributors are eligible to become team members after they have
continued to contribute actively for a year. Trusted contributors can contact
team leads or will be contacted by team leads to enquire about promotion to team
membership. Trusted contributors who are eligible for team membership do not
have to become team members.

Unlike trusted contributors, team members are expected to consider themselves
as *maintainers* of the compiler - put otherwise, to be invested in the quality
of the compiler codebase and overall health of the compiler team, independent
of their own projects. Team membership is primarily intended to recognise and
encourage participation in activities which are vital to the success of the
compiler team and broader project.

> Team members are expected to participate in the ongoing maintainance tasks

Reviewer 8: ```suggestion
Team members are expected to participate in the ongoing maintenance tasks
```

Author: Fixed!

that the compiler team is responsible for (with all of the expected caveats
for vacation time, mental health breaks, etc) - listed as
[responsibilities][responsibilities] below. However, not all contributors need
to participate in these responsibilities to an equal degree. Contributors should
participate in these tasks to the degree that they are able - volunteers are not
expected to participate as much as those employed to work on the compiler, for
example. It is the responsibility of the compiler team leads to ensure that the
ongoing maintenance tasks of the team can be completed sustainably.

Team members have all of the same permissions and access as trusted
contributors. Like trusted contributors, team members are considered members of
the Rust project as a whole and are expected to follow the spirit of the [Code
of Conduct][coc].

Like trusted contributors, after inactivity for longer than a year, a

> contributor will be moved to alumni status. Members who are no longer able to

Reviewer 7: Maybe this could say "member" here to remove the possible ambiguity with regular and trusted contributors?

```suggestion
team member will be moved to alumni status. Members who are no longer able to
```

      an FCP thoroughly. FCP reviewers should also consider reaching out to relevant
      domain experts and soliciting their opinions whenever possible. Any team or
      project member can raise concerns with an FCP, which will be considered by the
      FCP reviewers.
      
      To function effectively, it is recommended that there be 4 - 8 FCP
      reviewers at any time. If less than 4 FCP reviewers are available, the compiler
      team co-leads will act as FCP reviewers until the reviewers can be found.

- Performance triage
    - There is a rotation of team members and other project members who check
      all of the interesting performance benchmarks from the last week to produce a
      report summarizing the improvements and regressions. This is valuable to keep
      track of the compiler's performance over time and make sure that regressions are
      being addressed.

- Issue prioritisation
    - The compiler team has a prioritisation procedure and policy to identify
      and label issues according to their importance. These labels feed into the
      backport procedure (what's worth being backported) and work priorities of team
      members.

- Backport reviews

>     - On a regular basis, some team members participate a review of pull

Reviewer 7: ```suggestion
    - On a regular basis, some team members participate in a review of pull
```

Author: Fixed!

      other tasks requiring broader discussion.
      
      To establish a reasonable quorum of triage members, it is recommended that
      at least 4 members participate in triage meetings. In the event there are not
      enough triage members, the compiler team co-leads will act as triage members
      until additional members are found.
      
- Review rotation
    - Every week, lots of pull requests are submitted to the compiler which need
      to be reviewed. Being on the review rotation is one of the primary ways that
      team members can help keep the wheels turning in the compiler team.

      It is **strongly** encouraged that all team members be a part of the
      review rotation.

- Operations
    - There are various *operations* tasks like agenda preparation and taking
      meeting notes which are very useful for the team.

      This list isn't exhaustive, and this RFC shouldn't be considered the canonical
      list of these responsibilities. Similarly, this RFC isn't intended to define how
      these responsibilities are conducted (in meetings or asynchronously, etc), that
      should be decided and documented by those involved in each.

> While this RFC doesn't aim to be authorative with respect to how team members

Reviewer 8: ```suggestion
While this RFC doesn't aim to be authoritative with respect to how team members
```

Author: Fixed!

who take on additional responsibilities are recognised, one way would be for
team members who take on additional responsibilities to record this in the
*rust-lang/team* metadata (using the `roles` key), e.g.

```toml
[people]
members = [
    { github = "davidtwco", roles = ["compiler-backport"] }
]

[[roles]]
id = "compiler-backport"
description = "Backport Reviewer"
```

By tracking responsibility participation in the team repository, it is easier
for the team leads to have visibility into the participation in each to ensure
that it is sustainable. Participants will be recognised for the additional
responsibilities they participate in on the compiler team's page on the Rust
website.

Drawbacks
=========
[drawbacks]: #drawbacks

- Granting permissions earlier may be a risk
    - We haven't had any issues with contributors having staying power to the
      extent that we would trust them with permissions and then having those be used

>       inappropriately. We can always revert changes if neccessary.

Reviewer 8: ```suggestion
      inappropriately. We can always revert changes if necessary.
```

Author: Fixed!

- Expectations of team members
    - This RFC formally establishes expectations which come with team
      membership. Some team members already assume that these expectations are there,
      but this wasn't made explicit when current team members were made team members.

Rationale and alternatives
==========================
[rationale-and-alternatives]: #rationale-and-alternatives

- Get better at doing nominations
    - A lot of this proposal's simplification of the way that promotions are
      granted is based on the premise that our current system doesn't work well for us
      - but we could just try and do the current system better.

> - Only change review responsibilites

Reviewer 8: ```suggestion
- Only change review responsibilities
```

for whom they would be able to work more effectively with these permissions and
can be trusted to use them responsibly. Team members do not need to
have experience with most of the compiler, and can be specialised to specific
subsystems of the compiler.

Team members are granted *r+*, *try*, *triagebot*, *rustc-perf*, and
*crater* permissions; *organization membership*; and *dev desktop* access.
Team members are considered members of the Rust project as a whole, and
are automatically eligible for any benefits that incurs (e.g. invitations to
meetups of project members). As representatives of the Rust project, trusted
contributors are expected to obey not just the letter of the
[Code of Conduct][coc] but its spirit.

Team members can choose to take on additional responsibilities, such as
those listed in the [responsibilities][responsibilities] section. Participating
in the team's review rotation is encouraged.

If a team member becomes inactive (the contributor's prior contributions and
other interactions with the project cease) for longer than a year, the team
member will be moved into alumni status. At any point in the future, they can

> ask to be re-instated at the team member level if they desire.

Reviewer 8: As pointed out that inactive members can become active again at any time. So, would it make sense establlishing to **six months** the cutoff to alumni status?

The rationale being that an inactive membership could theoretically add a bit of drag esp. when voting or other assignments are involved. A shorter cutoff could help clarify sooner a situation and help tackling pending work (f.e. PRs or review rotation)

Reviewer 8: By the way, the current RFC says [6 months](https://rust-lang.github.io/rfcs/2689-compiler-team-contributors.html#automatic-alumni-status-after-6-months-of-inactivity) of inactivity

Reviewer 9: mostly vibe based, but I'd prefer to keep the cutoff at 6 months as a full year feels a bit long

Author: Changed to six months.

[Compiler Development Guide]: https://rustc-dev-guide.rust-lang.org/

Maintainers
-----------

> [maintainers]: #maintainers

Reviewer 8: I don't know where but would it make sense to point out somewhere that being maintainers is completely orthogonal to team [rolling leadership](https://rust-lang.github.io/rfcs/3262-compiler-team-rolling-leads.html#term-length-and-leader-selection)?

A maintainer can also never be a team lead and there is no requirement/expectations in this regard (IIUC).

Author: I've added a brief section on team leads pointing at that RFC.

contribute actively for a year. Team members can contact team leads or will
be contacted by team leads to enquire about promotion to maintainership. Team
members who are eligible for maintainership do not have to become maintainers.

Unlike team members, maintainers are expected to consider themselves
as exactly that, *maintainers*, of the compiler - put otherwise, to be invested
in the quality of the compiler codebase and overall health of the compiler
team, independent of their own projects. Maintainership is primarily intended
to recognise and encourage participation in activities which are vital to the
success of the compiler team and broader project.

Maintainers are expected to participate in the ongoing maintenance tasks
that the compiler team is responsible for (with all of the expected caveats
for vacation time, mental health breaks, etc) - listed as
[responsibilities][responsibilities] below. However, not all maintainers need
to participate in these responsibilities to an equal degree. Maintainers should
participate in these tasks to the degree that they are able - volunteers are not
expected to participate as much as those employed to work on the compiler, for

> example. It is the responsibility of the compiler team leads to ensure that the

Reviewer 3: I think there is an unstated consequence here that if maintainers are not actively participating in compiler team maintenance (minus stated caveats), they can be removed as maintainers and this is as the discretion of team leads.

Author: Added a little bit of wording for this, may want more, let me know what you think.

ongoing maintenance tasks of the team can be completed sustainably.

Maintainers have all of the same permissions and access as team members.
Like team members, maintainers are considered members of the Rust project as a
whole and are expected to follow the spirit of the [Code of Conduct][coc].

> Like team members, after inactivity for longer than a year, a

Author: Changed to six months.

> It probably goes without saying that it would be nice if maintainers could have the courtesy of giving advice if they're finding themselves becoming inactive for a long time?

I'm not quite sure what you mean by this.

> Like team members, after inactivity for longer than a year, a

Reviewer 8: sorry for my clumsy English :) I meant that since maintainers are or will be doing maintenaince work, perhaps serving for longer or more regularly than other contributors, It would probably be good if maintainers do not disappear out of the blue and give notice if they foresee such situation.

(just a side-thought not meant to be bolted into this RFC)

maintainer will be moved to alumni status. Maintainers who are no longer able to
help maintain the compiler but otherwise wish to continue contributing to the
compiler can also be moved to alumni status and retain their team member
status. Alumni can ask to be reinstated in the future.

[coc]: https://www.rust-lang.org/policies/code-of-conduct

Responsibilities
----------------
[responsibilities]: #responsibilities

There are various responsibilities that a maintainer could take on to help the

> team. All maintainers must participate in at least one activity.

Reviewer 6: Do you mean in the short term ("I'm currently sick"/"away on vacation"/"taking a temporary leave"/etc) or something else?

> team. All maintainers must participate in at least one activity.

Reviewer 3: Short term (things like you mentioned), but also long term - people who's primary job is do work on Rust should probably not just be capping themselves at one activity just because it's "the requirement". (But this is also where intangible things come into play; I can imagine that someone might be involved in FCPs but also mentoring someone, for example.)

> team. All maintainers must participate in at least one activity.

Reviewer 6: Ah, thanks for elaborating! Yes, I would hope that people don't feel they can *only* do one thing but at the same time, we should be very cautious about encouraging folks to overextend themselves.

> team. All maintainers must participate in at least one activity.

Reviewer 3: Yeah, it's a tough line to walk. It's maybe hard to put into the right words, but personally, I think maintainers should be thinking about more than just some subset of compiler or really just the code itself. I think maintainership implies that you're thinking about the _Project_, which includes present *and* future.

> team. All maintainers must participate in at least one activity.

Author: I've added some wording based on what you've discussed here. Let me know if it addresses what you were intending to convey.

> team. All maintainers must participate in at least one activity.

Reviewer 3: Looks good!

> Maintainers can get involved in any of these by contacting the team leads, by

Reviewer 3: It may also be worth mentioning meta things that involve the team itself - maintenance of tooling the team uses, participation in things like this RFC, mentorship for contributors and members, etc. We often forget about the intangible things, but they ultimately a big part of "maintainership".

      by those engaged in doing so and that the team's work isn't unnecessarily
      delayed, FCP reviewers who aren't doing this defeat the point of the
      responsibility existing rather than being something all maintainers do.
      Any reviewer removed can ask to be re-added when they have the bandwidth
      to participate in FCP reviews.

- Performance triage
    - There is a rotation of maintainers and other project members who check
      all of the interesting performance benchmarks from the last week to produce a
      report summarizing the improvements and regressions. This is valuable to keep
      track of the compiler's performance over time and make sure that regressions are
      being addressed.

- Issue prioritisation
    - The compiler team has a prioritisation procedure and policy to identify
      and label issues according to their importance. These labels feed into the
      backport procedure (what's worth being backported) and work priorities of
      maintainers.

- Backport reviews
    - On a regular basis, some maintainers participate in a review of pull
      requests which have been nominated for backporting to the beta or stable

>       release. This involves a judgement call on the risk of backporting a

Reviewer 10: My english is not the best but I think it's judgment, right?

```suggestion
      release. This involves a judgment call on the risk of backporting a
```

      until additional members are found.

- Review rotation
    - Every week, lots of pull requests are submitted to the compiler which need
      to be reviewed. Being on the review rotation is one of the primary ways that
      maintainers can help keep the wheels turning in the compiler team.

      It is **strongly** encouraged that all maintainers be a part of the
      review rotation.

- Operations
    - There are various *operations* tasks like agenda preparation and taking
      meeting notes which are very useful for the team.

      This list isn't exhaustive, and this RFC shouldn't be considered the canonical
      list of these responsibilities. Similarly, this RFC isn't intended to define how
      these responsibilities are conducted (in meetings or asynchronously, etc), that
      should be decided and documented by those involved in each.

While this RFC doesn't aim to be authoritative with respect to how maintainers
who take on additional responsibilities are recognised, one way would be for
maintainers who take on additional responsibilities to record this in the
*rust-lang/team* metadata (using the `roles` key), e.g.

```toml
[people]

> members = [

Reviewer 3: Responding to my above comment: having all members,l (maintainers or not) be in the compiler toml will make tracking roles easier too.

Author: I agree. I've removed the teams repo stuff from the RFC because I don't want to prescribe how we represent this all in that here.

      of the compiler but could be reasonable if implementing all of the changes
      described here will take too long.

Prior art
=========
[prior-art]: #prior-art

- [Responsibilities][responsibilities] are similar to [an unsubmitted proposal
  by Niko Matsakis in December 2020 to have "elected officers"][officers] within
  the compiler team responsible for different team functions. This RFC shares
  many of the goals of Niko's earlier proposal, but is slightly less formal -
  responsibilities are loosely-defined groups of contributors rather than elected
  positions, and there is no rotations or term limits.

  In this RFC's proposal, it is expected that responsibilities are shared
  amongst a group of team members, and that team members do less of other
  responsibilities so that their workload is sustainable, but this isn't enforced.
  Team leads are instead responsible for ensuring that the team is large enough to
  perform each responsibility sustainably.

[officers]: https://hackmd.io/S9xqmwJbSI-a9mFdK9yQBA

Unresolved questions
====================
[unresolved-questions]: #unresolved-questions

None!

> Future possibilities

Reviewer 3: I think another future possibility is further elaboration of some minimum requirement for maintainers.

Reviewer 8: Yes, I was also thinking to something along that line esp. regarding being a more stable presence in the Rust project (see [comment](https://github.com/rust-lang/rfcs/pull/3599/files#r1551099380)).

After reading this document, I infer that maintainers of the Rust compiler are expected to contribute more consistently than team members.

Author: Added some wording for this.

maintainers will be taken into account. It is important to note that many kinds
of contributions will be considered such as code contributions, helping with
issue triage and bisection, running meetings and creating minutes, documentation
contributions for rustc internals or the [Compiler Development Guide], etc.

Team member is a mix of RFC 2689's "working group participant" and
"compiler team contributor" roles. It is explicitly intended to be granted more
liberally to contributors who have demonstrated competence and trustworthiness,
for whom they would be able to work more effectively with these permissions and
can be trusted to use them responsibly. Team members do not need to
have experience with most of the compiler, and can be specialised to specific
subsystems of the compiler.

Team members are granted *r+*, *try*, *triagebot*, *rustc-perf*, and *crater*
permissions; *organization membership*; and *dev desktop* access. Team members
can second major change proposals. Team members are considered members of the
Rust project as a whole, and are automatically eligible for any benefits that
incurs (e.g. invitations to meetups of project members). As representatives of

> the Rust project, trusted contributors are expected to obey not just the letter

Reviewer 11: ```suggestion
the Rust project, team members are expected to obey not just the letter
```
?

Author: Good catch!

- Get better at doing nominations
    - A lot of this proposal's simplification of the way that promotions are
      granted is based on the premise that our current system doesn't work well for us
      - but we could just try and do the current system better.
- Only change review responsibilities
    - We could instead try to increase the number of reviewers on the review
      queue by just amending the current compiler team membership policy to include
      review rotation duty. This does not improve our ability to correctly promote
      contributors or recognize the ways individuals contribute to the maintenance
      of the compiler but could be reasonable if implementing all of the changes
      described here will take too long.

# Prior art
[prior-art]: #prior-art

- [Maintenance activities][maintenance-activities] are similar to [an
  unsubmitted proposal by Niko Matsakis in December 2020 to have "elected
  officers"][officers] within the compiler team responsible for different team
  functions. This RFC shares many of the goals of Niko's earlier proposal,
  but is slightly less formal - activities are loosely-defined groups of

>   contributors rather than elected positions, and there is no rotations or term

Reviewer 12: Minor grammar nitpick

```suggestion
  contributors rather than elected positions, and there are no rotations or term
```

Author: Fixed!

    builds are hosted on `static.rust-lang.org` where we would never want
    malicious code.
- **review rotation**
  - Contributors on the review rotation will be randomly assigned to new pull
    requests submitted to the compiler. Being on the review rotation is one of the
    best ways for contributors to help the compiler team and learn new parts of
    the compiler. Review capacity is one of the most important resources that the
    team has, as it enables our progress in the compiler's continued development
    and maintenance.
- **organization membership**
  - Contributors that are added to the *rust-lang/compiler* team in the GitHub
    organization can be assigned to issues/pull requests, modify labels, receive
    group mentions and receive a "Member" badge next to their name.
- **rustc-perf**
  - Contributors with permissions to use *rustc-perf* can request benchmarking
    of their pull requests (and pull requests they are reviewing). *rustc-perf*
    permissions are useful for regular contributors as it is common to need to
    request benchmarks from contributors with permissions. *rustc-perf* permissions

>     only make sense alongside *try* permissions.

Reviewer 13: What is the reason to have this ability be separate from `try`? Is it purely to give further flexibility in case it is necessary; or are there additional security implications to `@perfbot` that need to be taken into the account that `@bors try` doesn't need to worry about?

Author: Just flexibility

Contributors start without any particular privileges, permissions or
responsibilities and can contribute whatever they'd like. Contributors can
progress to [Team Members][team-members] and then [Maintainers][maintainers].

## Team Members
[team-members]: #team-members

Being able to grant permissions to trusted contributors quickly is beneficial to
enable them to contribute to the project more efficiently and review and approve
work of their collaborators.

Any contributor can request to become a team members by contacting the
compiler team's leads, or current maintainers and team members can
nominate a contributor. Team leads will check for a reasonable contribution
history, and will check if the current team have any serious concerns related to
contributor conduct (waiting approximately one week).

When evaluating a candidate's contribution history, length of time and
consistency of contributions and interactions with other contributors and
maintainers will be taken into account. It is important to note that many kinds
of contributions will be considered such as code contributions, helping with
issue triage and bisection, running meetings and creating minutes, documentation

> contributions for rustc internals or the [Compiler Development Guide], etc.

Reviewer 13: I understand that it is possible to easily attach an objective criteria to the code contributions (number of merged pull requests, size of them) after the fact. What tools does GitHub (or our infrastructure in general) provide to list actions pertaining to triage, bisection, etc? Do we anticipate those to remain an action that a current member would need to consciously remind themselves to notice?

Reviewer 8: I can probably partially answer to this.

Non-code contributions to the project are basically invisible i.e. hard to quantify and produce stats for (unless, well, you look at the # of comments on GH and Zulip but they give a partial view[^1]). We have tooling to facilitate this work (f.e. triagebot) and people like me are working to improve this tooling.

> Do we anticipate those to remain an action that a current member would need to consciously remind themselves to notice?

IIUC your question you're asking if people doing non-code work know what to do and where to put their fingers on. The answer is yes and no because organizational work is relatively non-structured but we are looking into improving this. 

---

[^1]: reminds me when Apple used to measure developer throughput by counting LOCs/week ([link](https://computerhistory.org/blog/macpaint-and-quickdraw-source-code/?key=macpaint-and-quickdraw-source-code))

      of the compiler but could be reasonable if implementing all of the changes
      described here will take too long.

# Prior art
[prior-art]: #prior-art

- [Maintenance activities][maintenance-activities] are similar to [an
  unsubmitted proposal by Niko Matsakis in December 2020 to have "elected
  officers"][officers] within the compiler team responsible for different team
  functions. This RFC shares many of the goals of Niko's earlier proposal,
  but is slightly less formal - activities are loosely-defined groups of
  contributors rather than elected positions, and there are no rotations or term
  limits.

  In this RFC's proposal, it is expected that activites are shared amongst a
  group of team members, and that team members do less of other activities
  so that their workload is sustainable, but this isn't enforced. Team leads are
  instead responsible for ensuring that the team is large enough to perform each
  activity sustainably.

[officers]: https://hackmd.io/S9xqmwJbSI-a9mFdK9yQBA

# Unresolved questions
[unresolved-questions]: #unresolved-questions

None!

# Future possibilities
[future-possibilities]: #future-possibilities

> - Maintenance activities ould be formalized further - see references in

Reviewer 14: ```suggestion
- Maintenance activities could be formalized further - see references in
```

Author: Fixed!
